x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceAggregatesWithNull.java

```diff
@@ -0,0 +1,106 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.optimizer.rules.logical;
+
+import org.elasticsearch.compute.data.Block;
+import org.elasticsearch.compute.data.BlockUtils;
+import org.elasticsearch.xpack.esql.core.expression.Alias;
+import org.elasticsearch.xpack.esql.core.expression.EmptyAttribute;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.expression.Expressions;
+import org.elasticsearch.xpack.esql.core.expression.Literal;
+import org.elasticsearch.xpack.esql.core.expression.NamedExpression;
+import org.elasticsearch.xpack.esql.expression.function.aggregate.AggregateFunction;
+import org.elasticsearch.xpack.esql.expression.function.aggregate.Count;
+import org.elasticsearch.xpack.esql.expression.function.aggregate.CountDistinct;
+import org.elasticsearch.xpack.esql.plan.logical.Aggregate;
+import org.elasticsearch.xpack.esql.plan.logical.Eval;
+import org.elasticsearch.xpack.esql.plan.logical.LogicalPlan;
+import org.elasticsearch.xpack.esql.plan.logical.Project;
+import org.elasticsearch.xpack.esql.plan.logical.local.LocalRelation;
+import org.elasticsearch.xpack.esql.plan.logical.local.LocalSupplier;
+import org.elasticsearch.xpack.esql.planner.PlannerUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Looks for any aggregates functions (max(field), min(field)) that act on a foldable-to-null expression.
+ * Except COUNT_DISTINCT and COUNT (which should return 0), all other aggregate functions should return null.
+ *
+ * This applies to eval x = null | stats max(x) but also max(null) or max(2 + null).
+ * All aggregate functions that are also nullable (COUNT_DISTINCT and COUNT are exceptions), will get a NULL
+ * field replacement by the FoldNull rule, COUNT_DISTINCT will benefit from PropagateEvalFoldables.
+ */
+public final class ReplaceAggregatesWithNull extends OptimizerRules.OptimizerRule<Aggregate> {
+
+    public ReplaceAggregatesWithNull() {
+        super(OptimizerRules.TransformDirection.UP);
+    }
+
+    @Override
+    protected LogicalPlan rule(Aggregate aggregate) {
+        var aggs = aggregate.aggregates();
+        List<NamedExpression> newAggs = new ArrayList<>(aggs.size());
+        boolean changed = false;
+        List<Alias> transientEval = new ArrayList<>();
+
+        for (NamedExpression agg : aggs) {
+            Expression e = Alias.unwrap(agg);
+            Object value = null;
+            boolean isNewAgg = false;
+            if (e instanceof AggregateFunction af && Expressions.isNull(af.field())) {
+                isNewAgg = true;
+                if (af instanceof CountDistinct || af instanceof Count) {
+                    value = 0L;
+                }
+            } else if (Expressions.isNull(e)) {// FoldNull can replace an entire aggregate function with a null Literal
```

Comment:
> Does it make sense for FoldNull to still work on Aggregates? Maybe it's not worth changing, but as FoldNull would generate a wrong plan without this new rule, it would:
> 
> Make FoldNull not generate wrong plans
> Make this rule simpler (Moving the conditional to FoldNull really)

`FoldNull` has the job of quickly transforming expressions into `null` based on the **`Nullability` characteristic** of that expression. This is this rule **sole job**, it's contained, simple and well defined.

`Nullability` is a "shortcut" to quickly evaluate expressions to `null`. In the specific case of aggregation functions there are two exceptions to this `Nullability` thing - `count` and `count_distinct` - which can never be `null` (and I discovered the bug where we didn't mark `CountDistinct` as such, now it's fixed). If I were to change `FoldNull`, I would complicate its code with many exceptions to a general, simple rule. Maintaining this rule from this point on will be more complicated.

The alternative is to have this "aggregations on nulls scenarios" in a self contained rule - `ReplaceAggregatesWithNull` - which:
- contains the code of dealing with `null`s in aggs in a single place. If `FoldNull` would change, then this code would be split in two places - harder to maintain.
- deals with the "effect" of **potentially** having `FoldNull` run before it and inserting a `null` there

Your suggestion about `FoldNull` not doing the replacement for aggs is a valid one and will work if implemented, but my main argument in favor of the solution in this PR is that of **self-containment**, **clear separation of rules' tasks** and **ease of maintenance**.

> A small step towards having rules not depend on each other (Or at least not generate invalid plans)

Regarding this dependency, yes, it exists. And the dependency between optimization rules is a reality that spawns from the same arguments I mentioned above -  **self-containment**, **clear separation of rules' tasks** and **ease of maintenance**.

Each rule has one job and has some conditions that need to be `true` for the rule to be active. These conditions can be "activated" by the effect of other rules' executions.
If all the rules in the optimizer **would be independent**, each rule would perform the tasks that now can be found in **other rules** and the **code complexity and duplication** would be a nightmare.

I dare to say that **rule dependencies being inexistent is a fantasy**, we cannot have an easy to maintain code if this would be a reality.

Reply:
> I dare to say that rule dependencies being inexistent is a fantasy, we cannot have an easy to maintain code if this would be a reality.

Yeah, I'm more into making a "best effort" in making that happen. Not that it should be that way everywhere. That is, having rules without dependencies as a default, and then documenting dependencies for those that need them.

Btw, I'm fine with the solution, thanks for explaining!

Reply:
> Yeah, I'm more into making a "best effort" in making that happen. Not that it should be that way everywhere. That is, having rules without dependencies as a default, and then documenting dependencies for those that need them.

Definitely. And this is actually the hardest part when it comes to these rules: finding the best place to make the change so that rules remain self-contained, their job is as well defined as it can be and the change just "fits" in the ecosystem of the other rules. Of course, if another rule does something it shouldn't, that rule should change and maybe move what it was doing to somewhere else.